lib/Sema/ConstraintSystem.cpp

```diff
@@ -2356,3 +2356,110 @@ Expr *constraints::simplifyLocatorToAnchor(ConstraintSystem &cs,
+                                           AccessKind::ReadWrite, DC);
+      return access.getKind() == AccessStrategy::Storage;
+    };
+
+    auto isValidBaseDecl = [&](ValueDecl *baseDecl) -> bool {
+      if (!isDirectlyAccessedStoredVar(baseDecl))
+        return false;
+
+      // The base decl must either be static or global in order for it to be
+      // non-ephemeral.
+      return baseDecl->isStatic() ||
+             baseDecl->getDeclContext()->isModuleScopeContext();
+    };
+
+    bool isMember = false;
+    while (true) {
+      subExpr = subExpr->getSemanticsProvidingExpr();
+
+      if (auto *fve = dyn_cast<ForceValueExpr>(subExpr)) {
+        subExpr = fve->getSubExpr();
+        continue;
+      }
+
+      // Look through load exprs. These can occur when projecting through a
+      // property on a class, which we want to support if the class is
+      // non-final.
+      if (auto *le = dyn_cast<LoadExpr>(subExpr)) {
+        subExpr = le->getSubExpr();
+        continue;
+      }
+
+      if (auto *tee = dyn_cast<TupleElementExpr>(subExpr)) {
+        isMember = true;
+        subExpr = tee->getBase();
+        continue;
+      }
+
+      if (auto *mre = dyn_cast<MemberRefExpr>(subExpr)) {
+        if (!isDirectlyAccessedStoredVar(mre->getMember().getDecl()))
+          return false;
+        isMember = true;
+        subExpr = mre->getBase();
+        continue;
+      }
+      // TODO: Handle UnresolvedDotExpr.
+      break;
+    }
+
+    // First check if the base expr is a metatype – if so, we're dealing with
+    // a static member, which is non-ephemeral.
+    if (isMember && simplifyType(getType(subExpr))->is<AnyMetatypeType>())
+      return true;
+
+    // Otherwise, check to see if we have a base decl.
+    if (auto *dre = dyn_cast<DeclRefExpr>(subExpr))
+      return isValidBaseDecl(dre->getDecl());
+
+    // TODO: Handle potentially overloaded decl bases.
+    return false;
+  }
+  case ConversionRestrictionKind::TupleToTuple:
+  case ConversionRestrictionKind::DeepEquality:
+  case ConversionRestrictionKind::Superclass:
+  case ConversionRestrictionKind::LValueToRValue:
+  case ConversionRestrictionKind::Existential:
+  case ConversionRestrictionKind::MetatypeToExistentialMetatype:
+  case ConversionRestrictionKind::ExistentialMetatypeToMetatype:
+  case ConversionRestrictionKind::ValueToOptional:
+  case ConversionRestrictionKind::OptionalToOptional:
+  case ConversionRestrictionKind::ClassMetatypeToAnyObject:
+  case ConversionRestrictionKind::ExistentialMetatypeToAnyObject:
+  case ConversionRestrictionKind::ProtocolMetatypeToProtocolClass:
+  case ConversionRestrictionKind::PointerToPointer:
+  case ConversionRestrictionKind::ArrayUpcast:
+  case ConversionRestrictionKind::DictionaryUpcast:
+  case ConversionRestrictionKind::SetUpcast:
+  case ConversionRestrictionKind::HashableToAnyHashable:
+  case ConversionRestrictionKind::CFTollFreeBridgeToObjC:
+  case ConversionRestrictionKind::ObjCTollFreeBridgeToCF:
```

Comment:
Should this be llvm_unreachable, then?

Reply:
In this patch it could be, although IMO it should it be reasonable for the function to take an arbitrary conversion and answer the question "is this something that we could pass to a `@_nonEphemeral` parameter?" – which is true for the rest of these conversions, as they're not things that `@_nonEphemeral` restricts.